File: ros_topic_py_ws/src/topic_py/scripts/listener.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String

def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.data)
    
def listener():
    rospy.init_node('listener', anonymous=True)#创建订阅者节点; 第二个参数：为节点生成唯一的名称

    rospy.Subscriber("chatter", String, callback)#订阅了chatter话题,类型是std_msgs.msgs.String。当接收到新消息时，callback函数被调用

    rospy.spin() #只是不让你的节点退出，不影响订阅者回调函数，因为它们有自己的线程。

if __name__ == '__main__':
    listener()

File: ros_topic_cpp_ws/src/topic_cpp/src/listener.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"

//回调函数，当有新消息到达chatter话题时它就会被调用
//回调函数不能做太多事情，否则会一直运行，导致下一个触发始终不能再来运行，所以回调函数要高效
void chatterCallback(const std_msgs::String::ConstPtr& msg) //std_msgs::String是消息的结构，ConstPtr常指针，msg是以常指针的形式指向std_msgs::String
{
  ROS_INFO("I heard: [%s]", msg->data.c_str());//打印 指针类型调用
}

int main(int argc, char **argv)
{
  setlocale(LC_ALL,""); //打印中文的时候避免乱码
  ros::init(argc, argv, "listener");//初始化ROS,节点名称为talker 

  ros::NodeHandle n;//为这个进程的节点创建句柄

  //通过主节点订阅chatter话题.每当有新消息到达时，ROS将调用chatterCallback()函数.第二个参数是队列大小
  ros::Subscriber sub = n.subscribe("chatter", 1000, chatterCallback);//创建订阅者

  //等待消息传来，消息没来就等待，消息来了就去执行回调函数。
  ros::spin();//启动了一个自循环，一直循环等待回调函数，不会到下面打return 0

  return 0;
}
